service/history/historyReplicator.go

```diff
@@ -122,6 +124,98 @@ func newHistoryReplicator(shard ShardContext, historyEngine *historyEngineImpl,
 	return replicator
 }
 
+func (r *historyReplicator) SyncActivity(ctx context.Context, request *h.SyncActivityRequest) (retError error) {
+	domainID := request.GetDomainId()
+	execution := workflow.WorkflowExecution{
+		WorkflowId: request.WorkflowId,
+		RunId:      request.RunId,
+	}
+
+	context, release, err := r.historyCache.getOrCreateWorkflowExecutionWithTimeout(ctx, domainID, execution)
+	if err != nil {
+		// for get workflow execution context, with valid run id
+		// err will not be of type EntityNotExistsError
+		return err
+	}
+	defer func() { release(retError) }()
+
+	msBuilder, err := context.loadWorkflowExecution()
+	if err != nil {
+		if _, ok := err.(*workflow.EntityNotExistsError); !ok {
+			return err
+		}
+
+		// this can happen if the workflow start event and this sync activity task are out of order
+		// or the target workflow is long gone
+		// the safe solution to this is to throw away the sync activity task
+		// or otherwise, worker attempt will exceeds limit and put this message to DLQ
+		return nil
+	}
+
+	if !msBuilder.IsWorkflowExecutionRunning() {
+		// perhaps conflict resolution force termination
+		return nil
+	}
+
+	version := request.GetVersion()
+	scheduleID := request.GetScheduledId()
+	if scheduleID >= msBuilder.GetNextEventID() {
+		if version < msBuilder.GetLastWriteVersion() {
+			// activity version < workflow last write version
+			// this can happen if target workflow has
+			return nil
+		}
+
+		// version >= last write version
+		// this can happen if out of order delivery heppens
+		return ErrRetrySyncActivity
+	}
+
+	ai, isRunning := msBuilder.GetActivityInfo(scheduleID)
+	if !isRunning {
+		// this should not retry, can be caused by out of order delivery
+		// since the activity is already finished
+		return nil
+	}
+
+	if ai.Version > request.GetVersion() {
+		// this should not retry, can be caused by failover or reset
+		return nil
+	}
+
+	lastHeartbeatTime := time.Unix(0, request.GetLastHeartbeatTime())
+	if ai.Version == request.GetVersion() {
+		if ai.Attempt > request.GetAttempt() {
+			// this should not retry, can be caused by failover or reset
+			return nil
+		}
+		if ai.Attempt == request.GetAttempt() {
+			if ai.LastHeartBeatUpdatedTime.After(lastHeartbeatTime) {
+				// this should not retry, can be caused by out of order delivery
+				return nil
+			}
+			// version equal & attempt equal & last heartbeat after existing heartbeat
+			// should update activity
+		}
+		// version equal & attempt larger then existing, should update activity
+	}
+	// version latger then existing, should update activity
+
+	ai.Version = request.GetVersion()
+	ai.ScheduledTime = time.Unix(0, request.GetScheduledTime())
+	ai.StartedID = request.GetStartedId()
+	ai.StartedTime = time.Unix(0, request.GetStartedTime())
+	ai.LastHeartBeatUpdatedTime = lastHeartbeatTime
+	ai.Details = request.GetDetails()
+
+	err = msBuilder.UpdateActivity(ai)
```

Comment:
Can we follow the same pattern like rest of replicate code base.  Basically expose and API on mutable state to ReplicationActivityInfo?